geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationEventQueueManager.java

```diff
@@ -103,36 +103,47 @@ void drain(final ClientProxyMembershipID clientProxyMembershipID,
           + " without synchronization");
     }
 
-    drainEventsReceivedWhileRegisteringClient(clientProxyMembershipID, cacheClientNotifier);
-
     ClientRegistrationEventQueue registrationEventQueue =
         registeringProxyEventQueues.get(clientProxyMembershipID);
 
-    registrationEventQueue.lockForDraining();
-    try {
-      if (logger.isDebugEnabled()) {
-        logger.debug("Draining remaining events from registration queue for client proxy "
-            + clientProxyMembershipID + " with synchronization");
-      }
+    /*
+     * It is possible that multiple threads are actively attempting to register a given client,
+     * so the registration event queue may have already been drained, processed, and removed from
+     * registeringProxyEventQueues. So in that case we would expect registrationEventQueue
+     * to be null.
+     */
+    if (registrationEventQueue != null) {
+      drainEventsReceivedWhileRegisteringClient(registrationEventQueue,
```

Comment:
so the first call to this doesn't need a Lock?

Reply:
It's an optimization.  So we drain as many events as we can without taking out the lock (therefore continuing to allow adds), then finally take out the lock and finish draining/remove the queue.